### ISSUE TITLE
False positive when using inline `as` operator

### ISSUE THREAD

**Issue by lucastheisen** [dotnet/roslyn-analyzers#5720]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.401](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

I get a false positive when using inline [`as` operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#as-operator) paired with a [Null-conditional `?.`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-) as a validation gate.

### Steps To Reproduce

I have a validation interface

```csharp
    public interface IValidatableConfigurationBinding
    {
        /// <summary>Thows an exception if not valid. This is intended
        /// to be called immediately after binding to allow the bound object
        /// to validate the properties that were set during binding as early
        /// as possible.</summary>
        void Validate();
    }
```

Which is an _optional fail fast feature_ that i can use during dependency injection startup:

```csharp
        (runner as IValidatableConfigurationBinding)?.Validate();
```

This way any implementation class that implements this fail fast feature will _fail fast_.  However, this results in a CA1508 finding.  It can be worked around by switching to two separate lines:

```csharp
        var validatable = runner as IValidatableConfigurationBinding;
        validatable?.Validate();
```

But I cant see a logical reason that would be, so assuming false positive here.

### Expected behavior

No CA1508 finding

### Actual behavior

CA1508

**Comment by AArnott** [comment]
Here is another repro case where `(a as b)?.c` generates a false positive:

```cs
 ExpressionSyntax invokedMethodName = memberAccessExpression?.Name ?? invocation.Expression as IdentifierNameSyntax ?? (invocation.Expression as MemberBindingExpressionSyntax)?.Name ?? invocation.Expression;
```

[Full context](https://github.com/microsoft/vs-threading/pull/962/commits/9075367c3a1c0fb5e8e2232cade32b28ddd9f0fd#diff-383ce70bb6cac0027ef6a6d3c330b11e07cb9e61ef6e4e93d1a3294f6e32255f)

This was with .NET SDK 6.0.100.

**Comment by mavasani** [comment]
Verified no repro with latest .NET 8 SDK.

**Comment by mavasani** [comment]
Actually, I got a minimal repro on latest bits, re-opening

```csharp
class A
{
    public object FieldA;
}

class B : A
{
    public object FieldB;
}

class C
{
    void M(A a)
    {
        var x = a.FieldA as C ?? (a.FieldA as B)?.FieldA;
    }
}
```

**Comment by mavasani** [comment]
> Actually, I got a minimal repro on latest bits

My bad. This repros on latest available .NET 8 preview SDK, but not on the latest bits in the main branch. I'll add a regression test and close out this issue.
